src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.CompilationTrackerState.cs

```diff
@@ -114,56 +88,43 @@ private sealed class InProgressState : CompilationTrackerState
                     // as such, we must still determine what the up to date generated docs are.
                     Contract.ThrowIfTrue(!IsFrozen && generatorInfo.DocumentsAreFinal);
 
                     this.IntermediateProjects = intermediateProjects;
-                    this.CompilationWithGeneratedDocuments = compilationWithGeneratedDocuments;
+                    this.StaleCompilationWithGeneratedDocuments = staleCompilationWithGeneratedDocuments;
                 }
-            }
 
-            /// <summary>
-            /// A built compilation for the tracker that contains the fully built DeclarationTable,
-            /// but may not have references initialized
-            /// </summary>
-            private sealed class AllSyntaxTreesParsedState : CompilationTrackerState
-            {
-                /// <summary>
-                /// The result of taking the original completed compilation that had generated documents and updating them by
-                /// apply the <see cref="CompilationAndGeneratorDriverTranslationAction" />; this is not a correct snapshot in that
-                /// the generators have not been rerun, but may be reusable if the generators are later found to give the
-                /// same output.
-                /// </summary>
-                public Compilation? CompilationWithGeneratedDocuments { get; }
-
-                public AllSyntaxTreesParsedState(
+                public static InProgressState Create(
                     bool isFrozen,
                     Compilation compilationWithoutGeneratedDocuments,
                     CompilationTrackerGeneratorInfo generatorInfo,
-                    Compilation? compilationWithGeneratedDocuments) : base(
-                        isFrozen, compilationWithoutGeneratedDocuments, generatorInfo)
+                    Compilation? staleCompilationWithGeneratedDocuments,
+                    ImmutableList<(ProjectState state, CompilationAndGeneratorDriverTranslationAction action)> intermediateProjects)
                 {
-                    // If we're not in the frozen state, the documents must *not* be final.  We're not a FinalState, and
-                    // as such, we must still determine what the up to date generated docs are.
-                    Contract.ThrowIfTrue(!IsFrozen && generatorInfo.DocumentsAreFinal);
+                    Contract.ThrowIfTrue(intermediateProjects is null);
+
+                    // If we're not frozen, transition back to the non-final state as we def want to rerun generators
+                    // for either of these non-final states.
+                    if (!isFrozen)
+                        generatorInfo = generatorInfo.WithDocumentsAreFinal(false);
 
-                    CompilationWithGeneratedDocuments = compilationWithGeneratedDocuments;
+                    return new InProgressState(isFrozen, compilationWithoutGeneratedDocuments, generatorInfo, staleCompilationWithGeneratedDocuments, intermediateProjects);
```

Comment:
function is much simpler now that we don't have to distinguish between INProgressState and ALlSyntaxTreesParsedState